File: leetcode/101-symmetric-tree.js
// https://leetcode.com/problems/symmetric-tree/ (top-interview-questions)


// Definition for a binary tree node.
function TreeNode(val, left, right) {
  this.val = (val===undefined ? 0 : val)
  this.left = (left===undefined ? null : left)
  this.right = (right===undefined ? null : right)
}

/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isSymmetric = function(root) {

  const helper = (lNode, rNode) => {
    if (lNode === null && rNode === null) {
      return true;
    }

    if (lNode === null || rNode === null) {
      return false;
    }

    return (lNode.val === rNode.val) && helper(lNode.left, rNode.right) && helper(lNode.right, rNode.left); 
  }

  return helper(root.left, root.right);
};

File: leetcode/101-symmetric-tree.py
# https://leetcode.com/problems/symmetric-tree/ (top-interview-questions)

from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:

        def isMirror(node1, node2):
            if node1 == None or node2 == None:
                return node1 == node2
            return (node1.val == node2.val) and isMirror(node1.left, node2.right) and isMirror(node1.right, node2.left)
        
        return isMirror(root.left, root.right)
